### ISSUE TITLE
Adding `entries` field type to user blueprint breaks CP

### ISSUE THREAD

**Issue by jelleroorda** [statamic/cms#2855]
## Bug Description
After I add the following field to my `blueprints/user.yaml`, the whole CP is broken.

```yaml
      -
        handle: allowed_buildings
        field:
          type: entries
          collections:
            - buildings
          listable: hidden
```

## How to Reproduce

Add an entries field to your user blueprint, then try to load any page.

## Extra Detail

![Screenshot 2020-11-17 at 18 32 01](https://user-images.githubusercontent.com/9193686/99425492-32de4400-2903-11eb-9176-f7d121883d9e.png)

## Environment

Statamic 3.0.20 Pro
Laravel 7.29.3
PHP 7.4.9
No addons installed
Fresh install from `statamic/statamic`

**Comment by jelleroorda** [comment]
Possibly related to #2826

**Comment by damcclean** [comment]
Yeah, it's likely the exact same issue. 

I've not had a look into the cause yet but my guess is that it's trying to look for a locale on the user and users don't have locales attached to them.

**Comment by jasonvarga** [comment]
I believe this was fixed in 3.0.21.

**Comment by jelleroorda** [comment]
@jasonvarga I just upgraded to 3.0.28 but now I'm getting a 502 bad gateway with my set up.

> 2020/11/19 09:12:58 [error] 10#10: *2474 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 192.168.32.1, server: application.test, request: "GET /cp/dashboard HTTP/1.1", upstream: "fastcgi://192.168.32.6:9000", host: "application.test"

Removing the entries field from my user configuration gets me back to a normal dashboard. 

If I had to guess it's probably the same issue that I tackeled before on the front-end when using the `to_json` modifier. The `to_json` modifier will augment your collection before casting it to json, which can lead into an infinite loop when you have a cross reference in your collections (so an entries field in collection A where you select an entry from collection B, and then also select this entry from Collection A in an entries field in collection B). Maybe better explained is below (snippet taken from [Discord](https://discord.com/channels/489818810157891584/489819906540568593/777550603916279808) when I encountered the issue):

> Apparently there is an infinite loop when json encoding a collection entries, because the to_json modifier will first augment the collection of entries before converting it to an array. This is pretty much expected functionality, and works pretty much always, except in my use case. I have two collections of entries, let’s call them offices and cars. The problem exists because the office has a field “cars” which lists all the cars that the office has. Because a car can belong to multiple offices, the “car” entry blueprint also has a field “primary_office” which is also an entry field where you can select an office. Now if I have an office page where I convert the office to json with the modifier, it will keep augmenting infinitely if one of the cars has a selected primary office. Office to augmented => cars to augmented => primary office to augmented => cars of this primary office to augmented => etc...
Hope some of you guys can learn from this so you don’t feel puzzle and waste a few hours of your day for some code “that worked a while ago” and while you didn’t change the template it suddenly stops working.

I think this could be fixed by specifying on the `to_json` modifier how many levels of 'augmenting' you want, and after that level to do a shallow augment instead (since if I remember correctly it will only contain things like `id`, `permalink` etc. but not entries fields). However, this is easier said then done, since you would have to add this leveling thing to the augmentation process as well.

The "simple" solution I used for fixing my infinite loop on the website was by simply only selecting the fields I need before casting it to json:

```php
$entries = Entry::query()
            ->where('collection', 'buildings')
            ->get()
            ->map(fn($entry) => $entry->selectedQueryColumns([
                'id',
                'title',
                'featured_image',
                'featured',
                'permalink'
            ]));
```

**Comment by damcclean** [comment]
I've re-opened this issue then, I'll try and remember to reproduce at some point today. If I can see the issue, I'll see if I can make a pull request similar to #2853.

**Comment by damcclean** [comment]
Confirmed. This is still causing issues. However, instead of an exception being thrown, the Control Panel goes blank. This is what the console contains:

![image](https://user-images.githubusercontent.com/19637309/99704858-ddd03880-2a90-11eb-8ad9-f0587f0be9ea.png)

**Comment by jelleroorda** [comment]
Keep this one open for now, I’ll try to update the issue soon with some new information / a workaround that I’ve implemented for now. I think that could be implemented as a fix unless I’m missing something.

**Comment by jelleroorda** [comment]
Alright, so let's start with the actual issue that ended up happening for me. Like mentioned in [the original post](https://github.com/statamic/cms/issues/2855#issue-744940890), I ended up with an infinite recursion issue [like mentioned in this reply](https://github.com/statamic/cms/issues/2855#issuecomment-730245049). I managed to find the place where Statamic is currently creating that infinite loop, which stems from the augmenting of the currently logged in user. Since deep augmentation can create an infinite loop I have overridden the `Statamic\Http\View\Composers\JavascriptComposer` which binds the user to the view. Instead of the default 'deep' augmentation I've instead used the 'shallow' augmentation which only augments some basic fields. So far I've not experienced any issues with this, but maybe @jasonvarga will know if this could cause any issues that I've not yet experienced in the CP.

The piece of code changed:
```php
    protected function user($user)
    {
        if (! $user) {
            return [];
        }

        return array_merge($user->toAugmentedArray(), [
            'preferences' => Preference::all(),
            'permissions' => $user->permissions()->all(),
        ]);
    }
```
to
```php
    protected function user($user)
    {
        if (! $user) {
            return [];
        }

        return array_merge($user->toShallowAugmentedArray(), [
            'preferences' => Preference::all(),
            'permissions' => $user->permissions()->all(),
        ]);
    }
```

So basically the concrete question is: **Can shallow augmenting of the user cause issues in the CP?** If we're pretty sure that it doesn't I can make a PR to make the above mentioned change (or you can just close it with a commit).

**Comment by jasonvarga** [comment]
For now at least, keep your view composer customization in place. Good workaround.
